### PR TITLE
Decode gates, quality telemetry, adaptive debounce

### PR DESCRIPTION
**Pull request by FuJacob**
## Summary

Stacked on #686. The decode-side "showing nothing beats showing garbage" gates, the local counters that prove whether they help, and a latency-keyed debounce. Consumes the engine's new decode-quality primitives (cotabbyinference#10, on middleware `main`).

1. **Confidence floor, shipped at -1.5.** The runtime returns a typed output (text, mean per-token log-probability, withheld flag) so a confidence-suppressed completion is attributed as `lowConfidence` instead of "the model produced nothing". The -1.5 default came from a nine-point eval sweep (table below). Enabling the floor turns on the per-token logprob work the energy pass gated off; that cost is priced and bounded (p50 +13ms), and `cotabbyConfidenceFloorOverride` adjusts or disables (-inf) without a rebuild.
2. **argmax-EOG early stop.** The decode loop stops the moment the raw distribution's most-likely next token is end-of-generation (engine computes it while the logits row is hot; cotabbyinference#10). At temperature 0.1 sampling is near-greedy, so the eval shows no delta; this is tail insurance for the cases where the dist sampler draws past the model's intended stop. `cotabbyArgmaxStopDisabled` switches it off.
3. **Quality telemetry (always on, counters only).** Generations, shown, withheld-by-reason histogram (normalizer, confidence floor, seam guard), and accepted suggestions (once per suggestion, so word-by-word Tab walks do not inflate the rate), persisted across restarts and shown in the Performance pane with a reset control. Acceptance rate over the suppression histogram is the on-device ground truth for whether any of this stack actually helps.
4. **Adaptive debounce**: 15/25/55ms keyed to the last generation latency, configured value as fallback. Snappier on fast machines, calmer on slow ones.

## Validation

```
xcodebuild build-for-testing ... CODE_SIGNING_ALLOWED=NO   # ** TEST BUILD SUCCEEDED **
xcodebuild test-without-building ...                       # FULL suite: 1128 tests, 1123 passed, 0 failed, 5 skipped (gated evals)
swiftlint lint --quiet <changed files>                     # exit 0
xcodegen generate                                          # pbxproj additions only
```

**Confidence-floor sweep** (eval, Gemma E2B Q6_K, 117 cases, fixed seed; floor read at runtime via the override key so the sweep is rebuild-free):

| floor | qualityScore | precisionWhenShown | coverage | wrongShowRate | mustShow misses |
|---|---|---|---|---|---|
| off / -4 / -3 / -2.5 / -2 | 0.734 | 0.780 | 0.866 | 0.188 | 0 |
| **-1.5 (shipped)** | **0.744** | **0.820** | **0.794** | **0.137** | **0** |
| -1.0 | 0.624 | 0.849 | 0.474 | 0.068 | 0 |
| -0.5 | 0.459 | 0.857 | 0.124 | 0.017 | 0 |
| -0.3 | 0.407 | 1.000 | 0.031 | 0.000 | 0 |

Floors at or below -2 never fire on this model at temperature 0.1 (the model is confident even about garbage); -1.5 is the unique point where the composite rose while wrong-shows fell 27% relative and no must-show case was lost. A -0.05 sanity floor suppressed 117/117, proving the wiring end to end. Latency at -1.5: p50 187→200ms, p95 1036→1124ms (the logprob passes; engine-side cost is two O(vocab) scans per token).

## Linked issues

Refs #486 (quality/context usage now visible in the Performance window), #546 (adaptive debounce stops thrashing slow machines).

## Risk / rollout notes

- `LlamaRuntimeGenerating.generate` now returns `LlamaGenerationOutput` instead of `String` (internal protocol; both conformers and the test fake updated).
- The floor turns per-token logprob computation back on for the llama path. If field reports blame it, `defaults write com.jacobfu.tabby cotabbyConfidenceFloorOverride -float -1e9` restores the old posture without a build; telemetry's `lowConfidence` count shows exactly how often the gate fires in real use.
- Quality counters write one small UserDefaults blob per suggestion event; no content, no timestamps beyond a first-recorded date.
- New `SuggestionQualityMetricsStore` carries `nonisolated deinit {}` for the known app-hosted-test double-free; without it the store's tests crash with SIGABRT in full-suite runs (reproduced, fixed, full suite green).

<h3>Greptile Summary</h3>

This PR adds four new features on top of #686: a confidence floor gate and argmax-EOG early-stop at the decode layer, always-on quality telemetry counters (generated / shown / withheld-by-reason / accepted), and a latency-keyed adaptive debounce (15/25/55 ms). It also adds surface-context conditioning for the base model and a host-font caret advance measurement that fixes the overlay slide accuracy.

- **Decode gates**: `LlamaRuntimeCore` breaks the token loop before appending when `argmax_is_eog`; `LlamaSuggestionEngine` routes confidence-suppressed completions to a `SuggestionNormalizationResult(.lowConfidence)` so the coordinator attributes them correctly instead of reading as empty-model output. `resolvedStopAtArgmaxEOG` now accepts injectable `UserDefaults`, closing the testability gap noted in the prior review.
- **Quality telemetry**: `SuggestionQualityMetricsStore` is a lean `@MainActor` `ObservableObject` persisting lifetime counters via a JSON UserDefaults blob; each coordinator gate records its own named suppression reason, and `SuggestionEngineRouter.recordQualityOutcome` is the single accounting point for engine-level outcomes.
- **Adaptive debounce + surface context**: `DebouncePolicy` selects a debounce window from the last observed generation latency (fallback to configured value before first data); `SurfaceContextComposer` produces a sanitized writing-surface descriptor that suppresses itself for code editors and terminals.

<h3>Confidence Score: 5/5</h3>

Safe to merge; no new defects introduced on any path changed by this PR.

All four features are cleanly isolated, covered by targeted tests, and validated against a full 1128-test suite. The argmax-EOG token is correctly excluded from both the generated text and the sumLogprob accumulation before the confidence average is computed. The resolvedStopAtArgmaxEOG injection gap called out in the previous review is fixed. The suppression accounting is structurally sound for results that pass through only one accounting layer; concerns about results that simultaneously hit an engine-level and a lifecycle gate were covered in the prior review and are pre-existing rather than newly introduced here.

No files require special attention; all changes are internally consistent and well-tested.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| Cotabby/Models/SuggestionQualityMetricsStore.swift | New always-on quality counter store (generated/shown/suppressed-by-reason/accepted), persisted via UserDefaults JSON; correct @MainActor isolation with the standard nonisolated-deinit workaround. |
| Cotabby/Services/Runtime/LlamaSuggestionEngine.swift | Adds confidence floor and argmax-EOG stop wiring; resolvedStopAtArgmaxEOG now accepts injectable UserDefaults (fixes the testability gap flagged in the previous review), and LlamaGenerationOutput carries suppression attribution correctly. |
| Cotabby/Services/Runtime/LlamaRuntimeCore.swift | Decode loop gains the argmax-EOG early stop (break before extractPiece so the discarded token is excluded from both generatedText and sumLogprob) and returns LlamaGenerationOutput with averaged logprob and confidence-suppressed flag. |
| Cotabby/App/Coordinators/SuggestionCoordinator+Prediction.swift | All coordinator early-exit gates now record their own suppression reasons; the empty-result gate guards against double-counting engine-attributed reasons; recordShown placed correctly at the single path where a generation becomes visible. |
| Cotabby/App/Coordinators/SuggestionCoordinator+Acceptance.swift | Acceptance path records accepted suggestion on first chunk; InsertedTextAdvance provides host-font caret measurement for overlay sliding; predictedCaretRect now takes fieldStyle for improved accuracy. |
| Cotabby/Services/Runtime/SuggestionEngineRouter.swift | Single recordQualityOutcome method handles generated + engine-attributed suppression accounting for all engine paths; qualityMetricsStore correctly injected. |
| Cotabby/Support/DebouncePolicy.swift | Simple latency-keyed debounce (15/25/55ms), nonisolated enum, clean guard on nil/zero latency; the configured value is explicitly the pre-data fallback per the documented design. |
| Cotabby/Support/SurfaceContextComposer.swift | New surface-context composer with clear omission-beats-noise invariants; sanitizedTitle suffix-stripping uses .anchored+.backwards correctly; domain extraction strips path/query/www; no prompt injection via controlChar+quote filtering. |
| Cotabby/Support/AppSurfaceClassifier.swift | Clean bundle-ID classifier; terminal check runs on the original case before lowercasing for prefix tables; integrated-terminal flag takes precedence per the comment. |
| Cotabby/Support/InsertedTextAdvance.swift | Host-font caret advance measurement; correct nil guards; falls back to system font with host point size when face name is unavailable, which is better than the fixed 14pt ghost-font fallback it replaces. |
| Cotabby/Services/Focus/SurfaceContextCache.swift | Session-scoped AX capture cache keyed on PID:elementID:focusSeq; @MainActor isolation with nonisolated-deinit workaround; caches negative results to avoid re-probing unresponsive hosts. |
| Cotabby/UI/Settings/Panes/PerformancePaneView.swift | New suggestion quality section with counters, acceptance rate, top-4 suppression reasons, and a Reset control; raw enum key display was flagged in the prior outside-diff comment. |
| Cotabby/Support/SuggestionOverlayStabilityGate.swift | isAwaitingPostInsertionSync short-circuit correctly placed after the text-change check but before geometry checks, preventing the pre-publish stale caret from moving the overlay. |
| CotabbyTests/SuggestionQualityMetricsStoreTests.swift | Good coverage: accumulation, acceptance rate, cross-instance persistence via isolated UserDefaults suite, and reset. Avoids process-global state. |
| CotabbyTests/LlamaDecodeGateDefaultsTests.swift | Covers the defaults-write escape hatches for both confidence floor and argmax-stop via isolated UserDefaults suite; resolvedStopAtArgmaxEOG is now testable because the previous injection gap was fixed. |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[Engine Result] --> R[Router: recordGenerated]
    R --> RS{suppressionReason?}
    RS -- yes --> RSR[recordSuppressed engine reason]
    RS -- no --> C[Coordinator]
    RSR --> C
    C --> SD{Stale drop?}
    SD -- yes --> SDR[recordSuppressed discardedStaleContext]
    SDR --> END1([Drop])
    SD -- no --> ER{result.text empty?}
    ER -- yes, no engine reason --> ERR[recordSuppressed emptyUnattributed]
    ERR --> END2([Drop])
    ER -- yes, has engine reason --> END2
    ER -- no --> ST{Selected text?}
    ST -- yes --> STR[recordSuppressed discardedSelection]
    STR --> END3([Drop])
    ST -- no --> AE{Stale accept echo?}
    AE -- yes --> AER[recordSuppressed discardedAcceptEcho]
    AER --> END4([Drop])
    AE -- no --> SG{Seam guard?}
    SG -- yes --> SGR[recordSuppressed seamMisspelling or seamJunkPunctuationRun]
    SGR --> END5([Drop])
    SG -- no --> SH[recordShown]
    SH --> SHOW([Show overlay])
    SHOW --> ACC{First Tab accept?}
    ACC -- yes, consumedCount == 0 --> ACCR[recordAcceptedSuggestion]
    ACC -- no --> ACCR
```

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `Cotabby/UI/Settings/Panes/PerformancePaneView.swift`, line 902-908 ([link](https://github.com/fujacob/cotabby/blob/c92d558a03689106b50ad9c16d2519314374aed8/Cotabby/UI/Settings/Panes/PerformancePaneView.swift#L902-L908)) 

   <a href="#">P2</a> **Raw enum key names surfaced verbatim in the UI**

   `topSuppressionReasons` maps the raw `suppressedByReason` dictionary keys directly to the displayed string. Those keys are `CompletionSuppressionReason` raw values (`"lowConfidence"`, `"emptyGeneration"`, `"seamMisspelling"`) plus coordinator-level strings (`"seamJunkPunctuationRun"`). A non-developer user reading the Performance pane would see `"lowConfidence 12, seamJunkPunctuationRun 3"` with no indication of what those mean. A small display-name map (or at least title-casing / spacing) would make the section usable outside of debugging contexts.

   Note: If this suggestion doesn't match your team's coding style, reply to this and let me know. I'll remember it for next time!

   <a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22fujacob%2Fcotabby%22%20on%20the%20existing%20branch%20%22quality%2Fdecode-gates-and-telemetry%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22quality%2Fdecode-gates-and-telemetry%22.%0A%0AThis%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20Cotabby%2FUI%2FSettings%2FPanes%2FPerformancePaneView.swift%0ALine%3A%20902-908%0A%0AComment%3A%0A**Raw%20enum%20key%20names%20surfaced%20verbatim%20in%20the%20UI**%0A%0A%60topSuppressionReasons%60%20maps%20the%20raw%20%60suppressedByReason%60%20dictionary%20keys%20directly%20to%20the%20displayed%20string.%20Those%20keys%20are%20%60CompletionSuppressionReason%60%20raw%20values%20%28%60%22lowConfidence%22%60%2C%20%60%22emptyGeneration%22%60%2C%20%60%22seamMisspelling%22%60%29%20plus%20coordinator-level%20strings%20%28%60%22seamJunkPunctuationRun%22%60%29.%20A%20non-developer%20user%20reading%20the%20Performance%20pane%20would%20see%20%60%22lowConfidence%2012%2C%20seamJunkPunctuationRun%203%22%60%20with%20no%20indication%20of%20what%20those%20mean.%20A%20small%20display-name%20map%20%28or%20at%20least%20title-casing%20%2F%20spacing%29%20would%20make%20the%20section%20usable%20outside%20of%20debugging%20contexts.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=fujacob%2Fcotabby&pr=688&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodex.svg?v=3">Fix in Codex</picture></a> <a href="https://app.greptile.com/ide/claude-code?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20Cotabby%2FUI%2FSettings%2FPanes%2FPerformancePaneView.swift%0ALine%3A%20902-908%0A%0AComment%3A%0A**Raw%20enum%20key%20names%20surfaced%20verbatim%20in%20the%20UI**%0A%0A%60topSuppressionReasons%60%20maps%20the%20raw%20%60suppressedByReason%60%20dictionary%20keys%20directly%20to%20the%20displayed%20string.%20Those%20keys%20are%20%60CompletionSuppressionReason%60%20raw%20values%20%28%60%22lowConfidence%22%60%2C%20%60%22emptyGeneration%22%60%2C%20%60%22seamMisspelling%22%60%29%20plus%20coordinator-level%20strings%20%28%60%22seamJunkPunctuationRun%22%60%29.%20A%20non-developer%20user%20reading%20the%20Performance%20pane%20would%20see%20%60%22lowConfidence%2012%2C%20seamJunkPunctuationRun%203%22%60%20with%20no%20indication%20of%20what%20those%20mean.%20A%20small%20display-name%20map%20%28or%20at%20least%20title-casing%20%2F%20spacing%29%20would%20make%20the%20section%20usable%20outside%20of%20debugging%20contexts.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=fujacob%2Fcotabby&pr=688&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaudeDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaude.svg?v=3">Fix in Claude Code</picture></a>
</details>

<sub>Reviews (4): Last reviewed commit: ["ci: retrigger checks after force-push sy..."](https://github.com/fujacob/cotabby/commit/98ca09794307efb43c126df09e6e8fd374589d49) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=37140919)</sub>

### CODE REVIEW

**Comment by FuJacob** [comment]
Merging on local CI-equivalent validation: GitHub Actions did not register runs for this branch's pushes (repo Actions is otherwise healthy; sibling PRs triggered normally), so the full suite was run locally on the exact head tree: build-for-testing succeeded, 1542 tests with 0 failures, SwiftLint clean, xcodegen drift-free. Review fixes included: the quality ledger now balances at every lifecycle exit, and the argmax toggle gained an injectable-defaults seam with tests.
